### PR TITLE
Feature parallel execution

### PR DESCRIPTION
**Pull request by MrM0nkey**
I implemented a framework to be able to execute experiments parallel. The number of processes is based on the machine power or more precisely the number of cpu cores. For this purpose I created the class [experimenter](https://github.com/nils-wisiol/pypuf/blob/8ff001ffdd35345c85a0f4c91e8fd3c0f81fbfbc/pypuf/experiments/experimenter.py) which handles all the parallel execution overhead. The experiments to be executet must implement the [experiment](https://github.com/nils-wisiol/pypuf/blob/8ff001ffdd35345c85a0f4c91e8fd3c0f81fbfbc/pypuf/experiments/experiment/base.py) base class. I have implemented prototype of a logistic regression learner [class](https://github.com/nils-wisiol/pypuf/blob/8ff001ffdd35345c85a0f4c91e8fd3c0f81fbfbc/pypuf/experiments/experiment/logistic_regression.py). Feedback especially to improve the logistic regression experiment is highly appreciated.

### CODE REVIEW

**Comment by MrM0nkey** [comment]
I recognized that executing the framework from terminal e.g. with sim_learn.py, the learner writes his status to stderr and messes up the log output. I would recommend to strip the output from the [learn](https://github.com/nils-wisiol/pypuf/blob/59f345eabe79973b4a4e24d0f0bc371225b75251/pypuf/learner/regression/logistic_regression.py#L209-L230) method for the moment, because I really want to avoid to inject code from parallel execution into other modules.  I would like to have a general logging style which is used by the parallel execution and the rest of the framework. A none generalized solution would be easier to write now but hard to change later. I would suggest to introduce a logging solution to the whole project which works with multiple processes in an new pull request or to remove the status informations from learner?

File: pypuf/experiments/experimenter.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
+import multiprocessing
+import logging
+from pypuf.tools import setup_logger
+
+# this semaphore should be only used by experiment.base.Experiment
+_semaphore = None
```
Comment (nils-wisiol) [inline pypuf/experiments/experimenter.py]: move this to a class variable to avoid global variables

File: pypuf/experiments/experimenter.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+# this queue should be only used by experiment.base.Experiment in oder to gather experiment output
+_queue = None
```
Comment (nils-wisiol) [inline pypuf/experiments/experimenter.py]: make this a class variable

File: pypuf/experiments/experimenter.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+
+def output_queue():
```
Comment (nils-wisiol) [inline pypuf/experiments/experimenter.py]: move this into `Experimenter`

File: pypuf/experiments/experimenter.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        It should protect from mess up the parallel execution.
+    :return: queue.Queue
+    """
+    global _queue
+    if _queue is None:
+        _queue = multiprocessing.Queue()
+    return _queue
+
+
+def semaphore():
```
Comment (nils-wisiol) [inline pypuf/experiments/experimenter.py]: move this into `Experimenter`

File: pypuf/experiments/experimenter.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        # wait for all processes to be finished
+        for job in jobs:
+            job.join()
+
+        # signal the output thread to be done
+        out_queue.put(Experimenter.FINISHED)
+
+    def run(self):
+        # master process
+        master_process = multiprocessing.Process(target=self.start_jobs, args=(self.out_queue,))
```
Comment (nils-wisiol) [inline pypuf/experiments/experimenter.py]: Why do we need `run()` and `start_jobs()` in different processes? Can we not just call `start_jobs()` directly?

File: pypuf/tools.py [hunk 6]
```diff
@@ -72,13 +76,40 @@ def compare_functions(x, y):
...
     def __init__(self, instance, N):
+        self.instance = instance
         self.challenges = array(list(sample_inputs(instance.n, N)))
         self.responses = instance.eval(self.challenges)
         self.N = N
+
+
+def setup_logger(logger_name, log_file="log", level=logging.INFO, write_file=True, write_console=True):
+    """
+        This function should be used to create a logger.
```
Comment (nils-wisiol) [inline pypuf/tools.py]: Better: "This function creates a logger, that can be used to ..."

File: pypuf/experiments/experimenter.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        # log the experiment output
+        while msg != Experimenter.FINISHED or out_queue.qsize() > 0:
+            # log the output
+            self.logger.info(msg+'\n')
+            msg = out_queue.get()
+
+    def run(self):
+        """
+            This method starts the logging and the execution of experiments.
+        :return:
```
Comment (nils-wisiol) [inline pypuf/experiments/experimenter.py]: no return value

File: pypuf/experiments/experimenter.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
+import multiprocessing
+import logging
+from pypuf.tools import setup_logger
+
+
+class Experimenter(object):
```
Comment (nils-wisiol) [inline pypuf/experiments/experimenter.py]: Add class doc

File: pypuf/experiments/experiment/base.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+class Experiment(object):
+    """
+        This class is the inteface for all experiments executed by the pypuf.experiments.experimenter.Experimenter class
+        and ensures compatibility.
+    """
+
+    def __init__(self, log_name, learner):
+        """
+        :param learner: This should be an instance of pypuf.simulation.base.Simulation
```
Comment (nils-wisiol) [inline pypuf/experiments/experiment/base.py]: "A pypuf.simulation.base.Simulation instance"

File: pypuf/experiments/experiment/base.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        :param log_name: This name should be unique
```
Comment (nils-wisiol) [inline pypuf/experiments/experiment/base.py]: "A unique name, used for"

File: pypuf/experiments/experiment/logistic_regression.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+from numpy import amin, amax, mean, array, append
+from pypuf.experiments.experiment.base import Experiment
+from pypuf.learner.regression.logistic_regression import LogisticRegression
+from pypuf.simulation.arbiter_based.ltfarray import LTFArray
+from pypuf import tools
+
+
+class ExperimentLogisticRegression(Experiment):
+    """
+        This class cloud be used to setup an experiment with logistic regression learner on a LTFArray puf simulation.
```
Comment (nils-wisiol) [inline pypuf/experiments/experiment/logistic_regression.py]: "This Experiment uses the logistic regression learner on an LTFArray PUF simulation."

File: pypuf/learner/regression/logistic_regression.py [hunk 12]
```diff
@@ -227,7 +226,4 @@ def learn(self):
             distance = (self.training_set.N - count_nonzero(self.training_set.responses == self.sign_combined_model_responses)) / self.training_set.N
             self.min_distance = min(distance, self.min_distance)
 
-        if not converged and distance > .01:
-            stderr.write('\rNOT CONVERGED                    ')
```
Comment (nils-wisiol) [inline pypuf/learner/regression/logistic_regression.py]: This is an important log message that -- I believe -- is not yet implemented in the new code. But we need it! :)

**Comment by nils-wisiol** [comment]
I like this PR a lot! Thanks for the great work.

Can you refactor `sim_learn.py` to benefit from the new Experiment class? That'd be great. Other than that and a few nitpicking comments, lgtm. :)

File: test/test_experimenter.py [hunk 13]
```diff
@@ -0,0 +1,21 @@
+import unittest
+from pypuf.simulation.arbiter_based.ltfarray import LTFArray
+from pypuf.experiments.experiment.logistic_regression import ExperimentLogisticRegression
+from pypuf.experiments.experimenter import Experimenter
+
+
+class TestExperimenter(unittest.TestCase):
+    def test_lr_experiments(self):
```
Comment (nils-wisiol) [inline test/test_experimenter.py]: Please user smaller parameter numbers here to decrease test runtime (currently 20s on my machine)

**Comment by MrM0nkey** [comment]
"Can you refactor sim_learn.py to benefit from the new Experiment class?" I will do this in a new pull request.

File: pypuf/learner/regression/logistic_regression.py (hunk 12 above)
Comment (MrM0nkey) [inline pypuf/learner/regression/logistic_regression.py]: I can see three options for this:

1. Use the logger from tools.py and probably mess up the multiprocessing log.
2. Set a flag as a member variable and check it when analysing the learner.
3. Create a nifty logging server which understands the current scopes of logging and handles requests correct.

What shall I do?

File: pypuf/experiments/experimenter.py (hunk 8 above)
Comment (MrM0nkey) [inline pypuf/experiments/experimenter.py]: Fixed that.

File: test/test_experimenter.py (hunk 13 above)
Comment (MrM0nkey) [inline test/test_experimenter.py]: Fixed that. It runs now for 4 secs on my machine.

File: pypuf/learner/regression/logistic_regression.py (hunk 12 above)
Comment (nils-wisiol) [inline pypuf/learner/regression/logistic_regression.py]: two sounds best to me

Comment (MrM0nkey) [inline pypuf/learner/regression/logistic_regression.py]: I have done solution two.

**Pull request merged**